src/StoreApi/Utilities/CartController.php

```diff
@@ -213,17 +215,130 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+
 		foreach ( $cart_items as $cart_item_key => $cart_item ) {
-			$this->validate_cart_item( $cart_item );
+			try {
+				$this->validate_cart_item( $cart_item );
+			} catch ( TooManyInCartException $error ) {
+				$too_many_in_cart_products[] = $error;
+			} catch ( NotPurchasableException $error ) {
+				$not_purchasable_products[] = $error;
+			} catch ( PartialOutOfStockException $error ) {
+				$partial_out_of_stock_products[] = $error;
+			} catch ( OutOfStockException $error ) {
+				$out_of_stock_products[] = $error;
+			}
+		}
+
+		$error = new WP_Error();
+
+		if ( count( $out_of_stock_products ) > 0 ) {
+			// translators: %s: product names.
+			$singular_error = __(
+				'%s is out of stock and cannot be purchased. It has been removed from your cart.',
+				'woo-gutenberg-products-block'
+			);
+			// translators: %s: product names.
+			$plural_error = __(
+				'%s are out of stock and cannot be purchased. They have been removed from your cart.',
+				'woo-gutenberg-products-block'
+			);
+
+			$error->add(
+				409,
+				$this->add_product_names_to_message( $singular_error, $plural_error, $out_of_stock_products )
+			);
+		}
+
+		if ( count( $not_purchasable_products ) > 0 ) {
+			// translators: %s: product names.
+			$singular_error = __(
+				'%s cannot be purchased. It has been removed from your cart.',
+				'woo-gutenberg-products-block'
+			);
+			// translators: %s: product names.
+			$plural_error = __(
+				'%s cannot be purchased. They have been removed from your cart.',
+				'woo-gutenberg-products-block'
+			);
+
+			$error->add(
+				409,
+				$this->add_product_names_to_message( $singular_error, $plural_error, $not_purchasable_products )
+			);
+		}
+
+		if ( count( $too_many_in_cart_products ) > 0 ) {
+			// translators: %s: product names.
+			$singular_error = __(
+				'There are too many %s in the cart. Only 1 can be purchased. The quantity in your cart has been reduced.',
+				'woo-gutenberg-products-block'
+			);
+			// translators: %s: product names.
+			$plural_error = __(
+				'There are too many %s in the cart. Only 1 of each can be purchased. The quantities in your cart have been reduced.',
+				'woo-gutenberg-products-block'
+			);
+
+			$error->add(
+				409,
+				$this->add_product_names_to_message( $singular_error, $plural_error, $too_many_in_cart_products )
+			);
+		}
+
+		if ( count( $partial_out_of_stock_products ) > 0 ) {
+			// translators: %s: product names.
+			$singular_error = __(
+				'There is not enough %s in stock. The quantity in your cart has been reduced.',
+				'woo-gutenberg-products-block'
+			);
+			// translators: %s: product names.
+			$plural_error = __(
+				'There are not enough %s in stock. The quantities in your cart have been reduced.',
+				'woo-gutenberg-products-block'
+			);
+
+			$error->add(
+				409,
+				$this->add_product_names_to_message( $singular_error, $plural_error, $partial_out_of_stock_products )
+			);
+
+			if ( $error->has_errors() ) {
```

Comment:
I think this is nested too deep? It's within the `if ( count( $partial_out_of_stock_products ) > 0 ) {` conditon.

Reply:
Good spot, I didn't see this when testing because I had partial out of stock products.